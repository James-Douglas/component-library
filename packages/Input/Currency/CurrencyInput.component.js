import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import { useId } from '@comparethemarketau/manor-hooks';
import { tooltipPropTypes } from '@comparethemarketau/manor-tooltip';
import Input, { getInitialValue } from '../Input.component';

const CurrencyInput = ({
  id: propsId,
  label,
  ariaLabelledBy,
  ariaDescribedBy,
  placeholder,
  value,
  prefillValue,
  handleChange,
  currencySymbol,
  required,
  disabled,
  readonly,
  autocomplete,
  tooltip,
  validationMessage,
  maxlength,
  className,
  handleFocus,
  handleBlur,
}) => {
  const id = useId(propsId);
  const [internalValue, setInternalValue] = useState(getInitialValue(value, prefillValue));

  const changeHandler = useCallback((val) => {
    // strip commas and leading zeros
    const newVal = Number(val.toString().replace(/[^0-9]+/g, '')).toString();
    setInternalValue(newVal);
    handleChange(Number(newVal));
  }, [setInternalValue, handleChange]);

  /**
   * Given the current value of the field, returns a mask array
   * e.g. for value "1234" returns [/\d/, ',', /\d/, /\d/, /\d/]
   * @param val - the current value of the input field (includes formatting!)
   * @returns {any[]} mask array
   */
  const maskFunc = useCallback((val) => {
    if (val === '0') return [];
    // strip commas
    const raw = val.toString().replace(/[^0-9]+/g, '');
    // convert to array & trim value if it exceeds maxlength
    const rawArray = raw.split('').splice(0, maxlength);
    const result = [];

    let count = 0;
    for (let i = rawArray.length; i > 0; i -= 1) {
      if (count === 3) {
        result.push(',');
        // reset counter
        count = 0;
      }
      count += 1;
      result.push(/\d/);
    }

    return result.reverse();
  }, [maxlength]);

  return (
    <Input
      id={id}
      label={label}
      ariaLabelledBy={ariaLabelledBy}
      ariaDescribedBy={ariaDescribedBy}
      type="tel"
      handleChange={changeHandler}
      prefixContent={currencySymbol}
      placeholder={placeholder}
      required={required}
      maxlength={maxlength}
      mask={maskFunc}
      value={internalValue}
      prefillValue={prefillValue}
      disabled={disabled}
      readonly={readonly}
      validationMessage={validationMessage}
      autocomplete={autocomplete}
      tooltip={tooltip}
      className={className}
      handleFocus={handleFocus}
      handleBlur={handleBlur}
    />
  );
};

CurrencyInput.propTypes = {
  /**
   *  Unique id for the Currency input
   */
  id: PropTypes.string,
  /**
   * Custom handler to attach to the input field - used to get the value of the field for example.
   */
  handleChange: PropTypes.func.isRequired,
  /**
   * Called on focus of the checkbox
   */
  handleFocus: PropTypes.func,
  /**
   * Called on blur of the checkbox
   */
  handleBlur: PropTypes.func,
  /**
   * Label for the input
   */
  label: PropTypes.oneOf([PropTypes.string, PropTypes.node]),
  /**
   * Space separated List of ids of elements used to label the component ( see this link for usage info https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute )
   */
  ariaLabelledBy: PropTypes.arrayOf(PropTypes.string),
  /**
   * Space separated List of ids of elements used to describe the component (tooltips etc) ( see this link for usage info https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute )
   */
  ariaDescribedBy: PropTypes.arrayOf(PropTypes.string),
  /**
   * Tooltip object (see Tooltip documentation)
   */
  tooltip: PropTypes.shape(tooltipPropTypes),
  /**
   * Displays given validation message and invalid styles on the component when provided.
   */
  validationMessage: PropTypes.string,
  /**
   * Prefills the input and applies browser autocomplete styles
   */
  prefillValue: PropTypes.string,
  /**
   * Sets the value of the input
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * The placeholder text for the input.
   */
  placeholder: PropTypes.string,
  /**
   * Adds/removes a supporting element, `<span>OPTIONAL</span>` to show the field is optional.
   */
  required: PropTypes.bool,
  /**
   * Disables the button via a class on its wrapper, and an attribute on the input.
   */
  disabled: PropTypes.bool,
  /**
   * Specifies that the currency input should be read-only.
   */
  readonly: PropTypes.bool,
  /**
   * Maximum length for the input element
   */
  maxlength: PropTypes.number,
  /**
   * Turn the browsers implementation of autocompletion/memory of forms on or off.
   */
  autocomplete: PropTypes.string,
  /**
   * Currency symbol to be displayed within the input prefix
   */
  currencySymbol: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  /**
   * Classes to be applied to the Currency component
   */
  className: PropTypes.string,
};

CurrencyInput.defaultProps = {
  id: null,
  maxlength: 15,
  value: null,
  tooltip: {},
  ariaLabelledBy: [],
  ariaDescribedBy: [],
  validationMessage: null,
  prefillValue: null,
  placeholder: '',
  autocomplete: 'off',
  currencySymbol: '',
  required: true,
  disabled: false,
  readonly: false,
  label: '',
  className: '',
  handleFocus: null,
  handleBlur: null,
};

export default CurrencyInput;
