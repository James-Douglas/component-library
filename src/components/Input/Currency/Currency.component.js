import React from 'react';
import PropTypes from 'prop-types';
import Input from 'components/Input/Input.component';

const CurrencyInput = ({
  id, label, placeholder, prefillValue, handleChange, currencySymbol, bordered, required, disabled, invalid, autocomplete, tooltip, maxlength,
}) => {
/* id, type, placeholder, prefillValue, required, disabled, bordered, invalid, prefixContent, suffixContent, label, tooltip, autocomplete, handleChange, valueMasking, */


  const valueMasking = (val) => {
    if (val === '') {
      return '';
    }

    let raw = val.toString().replace(/[^0-9]+/g, '');

    if (raw.length > maxlength) {
      raw = raw.substring(0, maxlength);
    }

    let parsed = parseInt(raw, 10);

    if (parsed) {
      parsed = parsed.toLocaleString();
    } else {
      parsed = raw;
    }

    return {
      raw,
      parsed,
    };
  };

  return (
    <>
      <Input
        id={id}
        label={label}
        type="tel"
        handleChange={handleChange}
        prefixContent={currencySymbol}
        placeholder={placeholder}
        bordered={bordered}
        required={required}
        maxlength={maxlength}
        valueMasking={valueMasking}
        prefillValue={prefillValue}
        disabled={disabled}
        invalid={invalid}
        autocomplete={autocomplete}
        tooltip={tooltip}
      />
    </>
  );
};

CurrencyInput.propTypes = {
  id: PropTypes.string.isRequired,
  handleChange: PropTypes.func.isRequired,
  label: PropTypes.string,
  prefillValue: PropTypes.string,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  bordered: PropTypes.bool,
  maxlength: PropTypes.number,
  invalid: PropTypes.bool,
  autocomplete: PropTypes.string,
  currencySymbol: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),

  tooltip: PropTypes.shape({
    title: PropTypes.string,
    body: PropTypes.string,
    boundingElementSelector: PropTypes.string,
    screenReaderLabel: PropTypes.string,
  }),
};

CurrencyInput.defaultProps = {
  maxlength: '15',
  prefillValue: '',
  placeholder: '',
  autocomplete: 'off',
  currencySymbol: '',
  required: true,
  disabled: false,
  bordered: true,
  invalid: false,
  label: '',
  tooltip: {},
};

export default CurrencyInput;
