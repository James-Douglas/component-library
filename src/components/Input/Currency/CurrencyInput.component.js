import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import Input from 'components/Input/Input.component';
import { tooltipPropTypes } from '../../Tooltip/Tooltip.component';
import { getInitialValue } from '../Input.component';
import useId from '../../../hooks/useId';

const CurrencyInput = ({
  id: propsId,
  label,
  placeholder,
  value,
  prefillValue,
  handleChange,
  currencySymbol,
  bordered,
  required,
  disabled,
  autocomplete,
  tooltip,
  validationMessage,
  maxlength,
  className,
  handleFocus,
  handleBlur,
}) => {
  const id = useId(propsId);
  const [internalValue, setInternalValue] = useState(getInitialValue(value, prefillValue));

  const changeHandler = useCallback((val) => {
    // strip commas and leading zeros
    const newVal = Number(val.toString().replace(/[^0-9]+/g, '')).toString();
    setInternalValue(newVal);
    handleChange(Number(newVal));
  }, [setInternalValue, handleChange]);

  /**
   * Given the current value of the field, returns a mask array
   * e.g. for value "1234" returns [/\d/, ',', /\d/, /\d/, /\d/]
   * @param val - the current value of the input field (includes formatting!)
   * @returns {any[]} mask array
   */
  const maskFunc = useCallback((val) => {
    if (val === '0') return [];
    // strip commas
    const raw = val.toString().replace(/[^0-9]+/g, '');
    // convert to array & trim value if it exceeds maxlength
    const rawArray = raw.split('').splice(0, maxlength);
    const result = [];

    let count = 0;
    for (let i = rawArray.length; i > 0; i -= 1) {
      if (count === 3) {
        result.push(',');
        // reset counter
        count = 0;
      }
      count += 1;
      result.push(/\d/);
    }

    return result.reverse();
  }, [maxlength]);

  return (
    <Input
      id={id}
      label={label}
      type="tel"
      handleChange={changeHandler}
      prefixContent={currencySymbol}
      placeholder={placeholder}
      bordered={bordered}
      required={required}
      maxlength={maxlength}
      mask={maskFunc}
      value={internalValue}
      prefillValue={prefillValue}
      disabled={disabled}
      validationMessage={validationMessage}
      autocomplete={autocomplete}
      tooltip={tooltip}
      className={className}
      handleFocus={handleFocus}
      handleBlur={handleBlur}
    />
  );
};

CurrencyInput.propTypes = {
  /**
   *  Unique id for the Currency input
   */
  id: PropTypes.string,
  /**
   * Custom handler to attach to the input field - used to get the value of the field for example.
   */
  handleChange: PropTypes.func.isRequired,
  /**
   * Called on focus of the checkbox
   */
  handleFocus: PropTypes.func,
  /**
   * Called on blur of the checkbox
   */
  handleBlur: PropTypes.func,
  /**
   * Label for the input
   */
  label: PropTypes.string,
  /**
   * Tooltip object (see Tooltip documentation)
   */
  tooltip: PropTypes.shape(tooltipPropTypes),
  /**
   * Displays given validation message and invalid styles on the component when provided.
   */
  validationMessage: PropTypes.string,
  /**
   * Prefills the input and applies browser autocomplete styles
   */
  prefillValue: PropTypes.string,
  /**
   * Sets the value of the input
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * The placeholder text for the input.
   */
  placeholder: PropTypes.string,
  /**
   * Adds/removes a supporting element, `<span>OPTIONAL</span>` to show the field is optional.
   */
  required: PropTypes.bool,
  /**
   * Disables the button via a class on its wrapper, and an attribute on the input.
   */
  disabled: PropTypes.bool,
  /**
   * The input field border style.
   */
  bordered: PropTypes.bool,
  /**
   * Maximum length for the input element
   */
  maxlength: PropTypes.number,
  /**
   * Turn the browsers implementation of autocompletion/memory of forms on or off.
   */
  autocomplete: PropTypes.string,
  /**
   * Currency symbol to be displayed within the input prefix
   */
  currencySymbol: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  /**
   * Classes to be applied to the Currency component
   */
  className: PropTypes.string,
};

CurrencyInput.defaultProps = {
  id: null,
  maxlength: 15,
  value: null,
  tooltip: {},
  validationMessage: null,
  prefillValue: null,
  placeholder: '',
  autocomplete: 'off',
  currencySymbol: '',
  required: true,
  disabled: false,
  bordered: true,
  label: '',
  className: '',
  handleFocus: null,
  handleBlur: null,
};

export default CurrencyInput;
