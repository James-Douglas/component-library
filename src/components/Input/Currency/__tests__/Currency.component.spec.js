import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Currency, { valueMasking } from '../Currency.component';

describe('valueMasking()', () => {
  it('returns a raw value and a parsed value', () => {
    const { raw, parsed } = valueMasking('2000', 15);

    expect(raw).toEqual('2000');
    expect(parsed).toEqual('2,000');
  });

  it('removes disallowed characters and formats the input', () => {
    const { raw, parsed } = valueMasking('2v222222p1', 15);

    expect(raw).toEqual('22222221');
    expect(parsed).toEqual('22,222,221');
  });

  it('does not allow leading zeros', () => {
    const { raw, parsed } = valueMasking('00000022222221', 15);

    expect(raw).toEqual('00000022222221');
    expect(parsed).toEqual('22,222,221');
  });

  it('limits input to given maxLength', () => {
    const { raw, parsed } = valueMasking('2222222222222222222222222', 15);

    expect(raw).toEqual('222222222222222');
    expect(parsed).toEqual('222,222,222,222,222');
  });
});

describe('Currency', () => {
  it('renders with minimal props', () => {
    const { container } = render(<Currency id="test-id" handleChange={() => {}} />);

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders with lots of props', () => {
    const { container } = render(
      <Currency
        id="test-id"
        label="label-test"
        placeholder="placeholder test"
        currencySymbol="$"
        maxlength={5}
        autocomplete="on"
        handleChange={() => {}}
        disableFieldset
      />,
    );

    const currencyInput = container.querySelector('input');
    const id = currencyInput.getAttribute('id');
    const label = container.querySelector('label');
    const placeholder = currencyInput.getAttribute('placeholder');
    const prefix = container.querySelector('.prefix');
    const autocomplete = currencyInput.getAttribute('autocomplete');

    expect(id).toBe('test-id');
    expect(label.textContent).toBe('label-test');
    expect(placeholder).toBe('placeholder test');
    expect(prefix.textContent).toBe('$');
    expect(autocomplete).toBe('on');
  });

  it('displays formatted input', () => {
    const { container } = render(
      <Currency
        id="test-id"
        disableFieldset
        handleChange={() => {}}
      />,
    );

    const currencyInput = container.querySelector('input');

    fireEvent.input(currencyInput, { target: { value: '2000000000' } });
    expect(currencyInput.value).toEqual('2,000,000,000');

    fireEvent.input(currencyInput, { target: { value: '200000000' } });
    expect(currencyInput.value).toEqual('200,000,000');

    fireEvent.input(currencyInput, { target: { value: '2000000' } });
    expect(currencyInput.value).toEqual('2,000,000');

    fireEvent.input(currencyInput, { target: { value: '200000' } });
    expect(currencyInput.value).toEqual('200,000');

    fireEvent.input(currencyInput, { target: { value: '20000' } });
    expect(currencyInput.value).toEqual('20,000');

    fireEvent.input(currencyInput, { target: { value: '2000' } });
    expect(currencyInput.value).toEqual('2,000');

    fireEvent.input(currencyInput, { target: { value: '200' } });
    expect(currencyInput.value).toEqual('200');

    fireEvent.input(currencyInput, { target: { value: '20' } });
    expect(currencyInput.value).toEqual('20');
  });

  it('fires an additional function supplied as a prop', () => {
    const handler = jest.fn();
    const { container } = render(
      <Currency
        id="test-id"
        handleChange={handler}
        disableFieldset
      />,
    );

    const currencyInput = container.querySelector('input');
    fireEvent.input(currencyInput, { target: { value: '2222222' } });

    expect(handler).toHaveBeenCalled();
    expect(handler.mock.calls.length).toBe(1);
    expect(handler.mock.calls[0][0]).toBe('2,222,222');
  });

  it('removes disallowed characters and formats the input', () => {
    const { container } = render(
      <Currency
        id="test-id"
        handleChange={() => {}}
        disableFieldset
      />,
    );

    const currencyInput = container.querySelector('input');
    fireEvent.input(currencyInput, { target: { value: '2v222222p1' } });

    expect(currencyInput.value).toEqual('22,222,221');
  });

  it('does not allow leading zeros', () => {
    const { container } = render(
      <Currency
        id="test-id"
        handleChange={() => {}}
        disableFieldset
      />,
    );

    const currencyInput = container.querySelector('input');
    fireEvent.input(currencyInput, { target: { value: '000000022' } });

    expect(currencyInput.value).toEqual('22');
  });

  it('limits input to given maxLength', () => {
    const { container } = render(
      <Currency
        id="test-id"
        handleChange={() => {}}
        maxlength={5}
        disableFieldset
      />,
    );

    const currencyInput = container.querySelector('input');
    fireEvent.input(currencyInput, { target: { value: '22222222' } });

    expect(currencyInput.value).toEqual('22,222');
  });
});
