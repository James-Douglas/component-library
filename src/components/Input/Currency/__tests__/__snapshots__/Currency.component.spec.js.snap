// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency renders with minimal props 1`] = `
"<div class=\\"sc-AykKH tctSj row sc-AykKJ duMRHg\\"><div class=\\"sc-fzXfLO hzGlg label\\"><div cols=\\"12\\" class=\\"sc-AykKI evPkqg 
        sc-AykKK gUjXDj
      \\"><label for=\\"test-id\\" class=\\"sc-fzXfLP fuIufG\\"></label></div></div></div><div class=\\"sc-AykKH tctSj row sc-fzXfLU gENaCA\\"><div cols=\\"10\\" class=\\"sc-AykKI emtNqC 
        sc-fzXfLV hMHPxM
      \\"><div class=\\"sc-fzXfMv eUXFYH input-container\\"><div class=\\"sc-fzXfMw guXMYH\\"><div class=\\"sc-fzXfMx fVMFMz input-clear-wrap\\"><input id=\\"test-id\\" name=\\"test-id\\" type=\\"tel\\" placeholder=\\"\\" autocomplete=\\"off\\" maxlength=\\"15\\" class=\\"sc-fzXfMy dNNhnu input-default \\" value=\\"\\"></div></div></div></div></div>"
`;
